internal/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,341 @@
+package selfupdate
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	wailsruntime "github.com/wailsapp/wails/v2/pkg/runtime"
+)
+
+// Version is the current version of the application. Set at compile time for production builds using ldflags (see tasks in the /tasks/build directory).
+var Version = "development"
+
+// noSelfUpdate is set to "true" for builds distributed to package managers to prevent auto-updating. It is typed as a string because the linker allows only setting string variables at compile time (see https://pkg.go.dev/cmd/link).
+// Set at compile time using ldflags (see the prod-noupdate task in the /tasks/build directory).
+var noSelfUpdate = "false"
+
+// releaseTrack is the release track to follow for updates. It currently only takes the value "stable".
+var releaseTrack = "stable"
+
+// manifestsBaseURL is the base URL for fetching update manifests.
+const manifestsBaseURL = "https://zenprivacy.net/update-manifests"
+
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type SelfUpdater struct {
+	version      string
+	noSelfUpdate bool
+	releaseTrack string
+	httpClient   HTTPClient
+}
+
+type Release struct {
+	Version     string `json:"version"`
+	Description string `json:"description"`
+	AssetURL    string `json:"assetURL"`
+	SHA256      string `json:"sha256"`
+}
+
+const (
+	appName = "Zen"
+)
+
+func NewSelfUpdater(httpClient HTTPClient) (*SelfUpdater, error) {
+	if httpClient == nil {
+		return nil, errors.New("httpClient is nil")
+	}
+
+	u := SelfUpdater{
+		version:      Version,
+		releaseTrack: releaseTrack,
+		httpClient:   httpClient,
+	}
+	switch noSelfUpdate {
+	case "true":
+		u.noSelfUpdate = true
+	case "false":
+	default:
+		return nil, fmt.Errorf("invalid noSelfUpdate value: %s", noSelfUpdate)
+	}
+
+	return &u, nil
+}
+
+func (su *SelfUpdater) checkForUpdates() (*Release, error) {
+	log.Println("checking for updates")
+	if su.noSelfUpdate {
+		log.Println("noSelfUpdate=true, self-update disabled")
+		return nil, nil
+	}
+	if su.version == "development" {
+		log.Println("version=development, self-update disabled")
+		return nil, nil
+	}
+
+	url := fmt.Sprintf("%s/%s/%s/%s/manifest.json", manifestsBaseURL, su.releaseTrack, runtime.GOOS, runtime.GOARCH)
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("create request: %w", err)
+	}
+
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("User-Agent", "zen-desktop")
+
+	res, err := su.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("do request: %w", err)
+	}
+
+	defer res.Body.Close()
+
```

Comment:
Go's http client doesn't return errors for responses with non-successful codes, so we should check it manually with something like
```go
if res.StatusCode != http.StatusOK {
  return nil, fmt.Errorf("manifest request failed with status code %d", res.StatusCode)
}
```